File: SockstoreNet/Application/UseCases/CreateProductUseCase.cs
using Application.Commands;
using Application.Domain;
using Application.Interfaces;
using Application.Ports;
using Vocabulary;

namespace Application.UseCases;

public class CreateProductUseCase(IProductPort port) : ICreateProduct
{
    public async Task<Product> Create(CreateProductCommand cmd)
    {
        var product = new ProductAggregate(
            new ProductId(),
            cmd.Name,
            cmd.Category,
            cmd.Price,
            cmd.Stock);

        await port.Save(product);
        return product.ToProduct();
    }
}

File: SockStoreJava/src/application/usecases/src/main/java/be/itenium/sockstore/usecases/CreateProductUseCase.java
package be.itenium.sockstore.usecases;

import be.itenium.sockstore.api.CreateProduct;
import be.itenium.sockstore.domain.ProductAggregate;
import be.itenium.sockstore.api.commands.CreateProductCommand;
import be.itenium.sockstore.domain.ProductPort;
import be.itenium.sockstore.vocabulary.Product;

public class CreateProductUseCase implements CreateProduct {
    private final ProductPort productPort;

    public CreateProductUseCase(ProductPort productPort) {
        this.productPort = productPort;
    }

    public Product create(CreateProductCommand command) {
        ProductAggregate nieuwProduct = new ProductAggregate(
                command.naam(),
                command.categorie(),
                command.prijs(),
                command.voorraad()
        );
        return productPort.save(nieuwProduct).toProduct();
    }
}